source/common/router/config_impl.cc

```diff
@@ -1257,42 +1318,29 @@ VirtualHostImpl::VirtualHostImpl(
     hedge_policy_ = virtual_host.hedge_policy();
   }
 
-  for (const auto& route : virtual_host.routes()) {
-    switch (route.match().path_specifier_case()) {
-    case envoy::config::route::v3::RouteMatch::PathSpecifierCase::kPrefix: {
-      routes_.emplace_back(new PrefixRouteEntryImpl(*this, route, optional_http_filters,
-                                                    factory_context, validator));
-      break;
-    }
-    case envoy::config::route::v3::RouteMatch::PathSpecifierCase::kPath: {
-      routes_.emplace_back(
-          new PathRouteEntryImpl(*this, route, optional_http_filters, factory_context, validator));
-      break;
-    }
-    case envoy::config::route::v3::RouteMatch::PathSpecifierCase::kHiddenEnvoyDeprecatedRegex:
-    case envoy::config::route::v3::RouteMatch::PathSpecifierCase::kSafeRegex: {
-      routes_.emplace_back(
-          new RegexRouteEntryImpl(*this, route, optional_http_filters, factory_context, validator));
-      break;
-    }
-    case envoy::config::route::v3::RouteMatch::PathSpecifierCase::kConnectMatcher: {
-      routes_.emplace_back(new ConnectRouteEntryImpl(*this, route, optional_http_filters,
-                                                     factory_context, validator));
-      break;
-    }
-    default:
-      NOT_REACHED_GCOVR_EXCL_LINE;
-    }
+  if (virtual_host.has_matcher() &&
+      !Runtime::runtimeFeatureEnabled("envoy.reloadable_features.experimental_matching_api")) {
+    throw EnvoyException("Experimental matching API is not enabled");
+  }
 
-    if (validation_clusters.has_value()) {
-      routes_.back()->validateClusters(*validation_clusters);
-      for (const auto& shadow_policy : routes_.back()->shadowPolicies()) {
-        ASSERT(!shadow_policy->cluster().empty());
-        if (!validation_clusters->hasCluster(shadow_policy->cluster())) {
-          throw EnvoyException(
-              fmt::format("route: unknown shadow cluster '{}'", shadow_policy->cluster()));
-        }
+  if (virtual_host.has_matcher() && !virtual_host.routes().empty()) {
+    throw EnvoyException("cannot set both matcher and routes on virtual host");
+  }
+
+  if (virtual_host.has_matcher()) {
+    RouteActionContext context{*this, optional_http_filters, factory_context};
+    RouteActionValidationVisitor validation_visitor;
+    Matcher::MatchTreeFactory<Http::HttpMatchingData, RouteActionContext> factory(
+        context, factory_context, validation_visitor);
+
+    matcher_ = factory.create(virtual_host.matcher())();
+  } else {
+    for (const auto& route : virtual_host.routes()) {
+      if (!route.has_match()) {
+        throw EnvoyException("RouteValidationError.Match");
```

Comment:
What does this error mean? Can you make it more descriptive and also maybe add more comments here?

Reply:
I actually don't think we need this any more after we moved to evaluating the resolved Route, let me just remove this